arch/x86_64/src/common/x86_64_syscall.c

```diff
@@ -0,0 +1,335 @@
+          /* If we are signalling a user process, then we must be operating
+           * on the kernel stack now.  We need to switch back to the user
+           * stack before dispatching the signal handler to the user code.
+           * The existence of an allocated kernel stack is sufficient
+           * information to make this decision.
+           */
+
+          if (rtcb->xcp.kstack != NULL)
+            {
+              uint64_t usp;
+
+              /* Copy "info" into user stack */
+
+              usp = rtcb->xcp.saved_ursp - 8;
+
+              /* Create a frame for info and copy the kernel info */
+
+              usp = usp - sizeof(siginfo_t);
+              memcpy((void *)usp, (void *)regs[REG_RSI], sizeof(siginfo_t));
+
+              /* Now set the updated SP and user copy of "info" to RSI */
+
+              regs[REG_RSP] = usp;
+              regs[REG_RSI] = usp;
+            }
+#endif
+
+          break;
+        }
+
+      /* cmd=SYS_signal_handler_return:  This a user signal handler callback
+       *
+       *   void signal_handler_return(void);
+       *
+       * At this point, the following values are saved in context:
+       *
+       *   cmd = SYS_signal_handler_return
+       */
+
+      case SYS_signal_handler_return:
+        {
+          /* Set up to return to the user-space. We need:
+           *
+           *   RCX = taskentry (SYSRETQ return address)
+           *
+           */
+
+          struct tcb_s *rtcb = nxsched_self();
+
+          /* Set up to return to the kernel-mode signal dispatching logic. */
+
+          DEBUGASSERT(rtcb->xcp.sigreturn != 0);
+
+          regs[REG_RCX]       = rtcb->xcp.sigreturn;
+          regs[REG_RSP]       = rtcb->xcp.saved_rsp;
+          rtcb->xcp.sigreturn = 0;
+
+          /* For kernel mode, we should be already on a correct kernel stack
+           * which was recovered in x86_64_syscall_entry.
+           */
+
+          break;
+        }
+#endif  /* CONFIG_BUILD_KERNEL */
+
+      /* This is not an architecture-specific system call.  If NuttX is
+       * built as a standalone kernel with a system call interface, then
+       * all of the additional system calls must be handled as in the
+       * default case.
+       */
+
+      default:
+        {
+          int             nbr  = cmd - CONFIG_SYS_RESERVED;
+          struct tcb_s   *rtcb = nxsched_self();
+          syscall_stub_t  stub = (syscall_stub_t)g_stublookup[nbr];
+
+          DEBUGASSERT(nbr < SYS_nsyscalls);
+          DEBUGASSERT(rtcb->xcp.nsyscalls < CONFIG_SYS_NNEST);
+
+          /* Setup nested syscall */
+
+          rtcb->xcp.syscall[rtcb->xcp.nsyscalls].sysreturn = regs[REG_RCX];
+          rtcb->xcp.nsyscalls += 1;
+
+#ifdef CONFIG_ARCH_KERNEL_STACK
+          /* Store reference to user RSP for signals */
+
+          rtcb->xcp.saved_ursp = regs[REG_RSP];
+#endif
+
+          /* Call syscall function */
+
+          ret = stub(nbr, arg1, arg2, arg3, arg4, arg5, arg6);
```

Comment:
What happens if the process raises a signal to itself during the handler ?

Reply:
The way it works on other platforms is that signal delivery is deferred during the execution of the system call. After the system call is done, the masked signals are released.

Not sure if ostest checks for this specific case, ostest kernel build support is relatively new. It is also a bit limited due to missing task_spawn() in kernel build.

I don't remember why the signals are deferred during syscall though, I'd have to dive into the code itself. I was just wondering why this logic was removed:
https://github.com/apache/nuttx/blob/37a0445ddbe280ccc9f11c68de2db1120531d150/arch/arm/src/armv7-a/arm_syscall.c#L533
And in syscall return:
https://github.com/apache/nuttx/blob/37a0445ddbe280ccc9f11c68de2db1120531d150/arch/arm/src/armv7-a/arm_syscall.c#L253-L254

Same in riscv / arm64:
https://github.com/apache/nuttx/blob/37a0445ddbe280ccc9f11c68de2db1120531d150/arch/risc-v/src/common/riscv_swint.c#L113-L133

Reply:
I have no idea why I missed this logic, but I think you're right. It seems that ostest doesn't test this case

Reply:
unfortunately with this change ostest crashes, so I'm reverting it for now

Reply:
I need to check if this deferral is even relevant nowadays. 

It might be a relic from the original system call implementation which did not execute the system call directly here, but rather dropped to thread mode via the exception return path and then invoked a reserved system call "return from syscall" to get back. Basically running the exception entry / exit logic twice (costing 2x more CPU for every system call...)

The old logic is removed in risc-v and arm64 but is still present in the original arm implementation.

Reply:
Btw, do you enable interrupts before you run the system call function here ? If not, then that is an error. System calls must run with interrupts enabled.

Reply:
> Btw, do you enable interrupts before you run the system call function here ? If not, then that is an error. System calls must run with interrupts enabled.

I just fixed it